Enable tracking of read status of topics

amickan opened the pull request:
Part of https://github.com/DIAGNijmegen/rse-roadmap/issues/407

amickan commented on app/grandchallenge/discussion_forums/models.py:
```diff
@@ -230,6 +240,11 @@ def save(self, *args, **kwargs):
 
         if adding:
             self.assign_permissions()
+            # mark topic (and hence this post) as read by the user
+            TopicReadRecord.objects.update_or_create(
+                user=self.creator,
+                topic=self.topic,
+            )
```
I'm wondering if we should do a permission check that `self.creator` has view permission for the topic. Where would be the best place for such a check? In the `save` method of the `TopicReadRecord` model? 
We're currently creating `TopicReadRecord`s here and in the view - in the view, we are already checking the permission.

amickan commented on app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py:
```diff
@@ -21,6 +22,36 @@ def migrate_challenge_forums(apps, schema_editor):
     for challenge in Challenge.objects.all():
         new_forum = Forum.objects.create()
+
+        # assign permissions
```
I just realized that this does not work as is actually. I need to retrieve the historical model records for the groups to get this to work. No big deal, but I'm thinking that it actually is easier and clearer to move this to the Forum model. Would you agree, James?

amickan replied:
Hm, on second thought that does not save us from also adding the permissions in the migration since the assign_permissions function would not be available to the migration code and would also not be triggered since the custom save() method is not avaialble either. Ok, never mind then, I'll update the migration here to use the appropriate historical Group model and leave the rest as is.

jmsmkn replied:
🦆 happy to help.

amickan replied:
Learning a lot about migrations today: `assign_perm` does not work in migrations (https://github.com/django-guardian/django-guardian/issues/281#issuecomment-156264129).

amickan replied:
I think I got now, thanks 🦆 James ;-)

jmsmkn commented on app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py:
```diff
@@ -44,11 +76,47 @@ def migrate_challenge_forums(apps, schema_editor):
+        "discussion_forums", "ForumTopic"
+    )
+
+    def get_matching_forum(*, old_forum_id):
+        old_forum = MachinaForum.objects.get(pk=old_forum_id)
+        return old_forum.challenge.discussion_forum
+
+    def get_matching_topic(*, old_topic_id):
+        old_topic = MachinaTopic.objects.get(pk=old_topic_id)
+        new_forum = get_matching_forum(old_forum_id=old_topic.forum.pk)
+        return ForumTopic.objects.get(
+            forum=new_forum,
+            creator=old_topic.poster,
+            subject=old_topic.subject,
+        )
```
Maybe create a function that takes in the models as arguments?

jmsmkn commented on app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py:
```diff
@@ -44,11 +45,47 @@ def migrate_challenge_forums(apps, schema_editor):
+            new_topic = get_matching_topic(old_topic_id=track.topic.pk)
+        except ObjectDoesNotExist:
+            continue
+        TopicReadRecord.objects.create(topic=new_topic, user=track.user)
```
Again, this is going to be a huge number of objects, so the writes also need to be chunked.

jmsmkn commented on app/grandchallenge/discussion_forums/templatetags/discussion_forum_extras.py:
```diff
@@ -0,0 +1,8 @@
+
+@register.simple_tag()
+def get_first_unread_topic_post_for_user(*, topic, user):
+    return topic.get_unread_topic_posts_for_user(user=user).first()
```
N+1 problem - this is going to perform one or two database queries per topic, one to get the `TopicReadRecord`, the other to get the post for that topic.

amickan replied:
True. I had not considered that yet. I'm only using this to scroll to the first unread post on the topic detail page. I could move that logic to the topic detail view, check there for unread posts for the user for just the topic in question (thus avoiding the N+1) and then redirect to the correct page accordingly. 

While looking at the queries, I also realized that we are wasting a lot of queries on getting the `user_profile_link` for the creator of the last post. The last post is currently a property on the Topic model, so we can't prefetch the user profile and verification. In the original machina implementation, last_post if a FK field that is updated whenever a new post is created. Having it as a model field would enable us to prefetch it accordingly, which I think might be necessary given that we call this twice per topic on the topic list view.

jmsmkn commented on app/grandchallenge/discussion_forums/views.py:
```diff
@@ -124,10 +125,24 @@ def get_context_data(self, **kwargs):
+
+        # mark topic as read by the user
+        TopicReadRecord.objects.update_or_create(
+            user=self.request.user,
```
Shouldn't the anonymous user be excluded?

amickan replied:
The anonymous user does not have permission to access this view, and so `self.request.user` would never be the anonymous user here. I can add an explicit check, of course, but I think it's not strictly necessary.

amickan replied:
I have added the extra check just in case

jmsmkn reviewed (commented):
Sorry I was so late with reviewing this. Please can you add some tests for the number of database queries for the list view to ensure that there is not an N+1 problem and that we are not tracking what the anonymous user has or has not read.

amickan commented on app/grandchallenge/discussion_forums/migrations/0002_migrate_old_forums.py:
```diff
@@ -44,11 +45,47 @@ def migrate_challenge_forums(apps, schema_editor):
                 )
 
 
+def migrate_topic_tracks(apps, schema_editor):
```
That could work. Another option would be to use `bulk_create` which should not trigger `auto_now_add`. I still have to test if that is how it works, but if so, that would have my preference.

jmsmkn replied:
Good to know, thanks for confirming that.

amickan commented on app/grandchallenge/discussion_forums/models.py:
```diff
@@ -193,6 +194,13 @@ def last_page_num(self):
+    def get_unread_topic_posts_for_user(self, *, user):
+        try:
+            read_record = TopicReadRecord.objects.get(user=user, topic=self)
+            return self.posts.exclude(created__lt=read_record.modified)
```
Same as above, I will need to unset auto_now

amickan commented:
> Sorry I was so late with reviewing this. Please can you add some tests for the number of database queries for the list view to ensure that there is not an N+1 problem and that we are not tracking what the anonymous user has or has not read.

I have addressed all comments, and added / updated tests, and this is ready for re-review. The only test I'm struggling with is the one that checks the number of queries. When using the debug toolbar, I see that with each new topic there is 1 additional query (to get the TopicReadRecord), but in the test it doesn't matter how many extra topics I add ,the number of queries stays the same. I'm probably doing something wrong, but I can't see at the moment. I'll check again tomorrow.

amickan commented on app/grandchallenge/discussion_forums/models.py:
```diff
@@ -342,3 +385,31 @@ class ForumPostUserObjectPermission(UserObjectPermissionBase):
+            "topic",
+        ]
+
+    def save(self, *args, **kwargs):
+        if self.user.username == settings.ANONYMOUS_USER_NAME or isinstance(
+            self.user, AnonymousUser
+        ):
```
Overall it does feel a little overkill to check this on the model given that we also have guardrails on the view and in the method that is used to create these instances.

jmsmkn replied:
How do you know in the future that things will be the same? We have these sorts of checks everywhere, like when adding users to groups.

amickan replied:
True. The check here is a good addition then.

amickan commented:
I think I have found a way to reduce the number of queries even further by doing some custom prefetching and then using those prefetched records in the template tag.

jmsmkn reviewed (approved):
LGTM, nice use of prefetch.
